presto-main/src/main/java/com/facebook/presto/cost/SimpleFilterProjectSemiJoinStatsRule.java

```diff
@@ -0,0 +1,162 @@
+            semiJoinNode = (SemiJoinNode) filterNode.getSource();
+        }
+
+        return calculate(filterNode, semiJoinNode, lookup, session, types);
+    }
+
+    private Optional<PlanNodeStatsEstimate> calculate(FilterNode filterNode, SemiJoinNode semiJoinNode, Lookup lookup, Session session, Map<Symbol, Type> types)
+    {
+        PlanNodeStatsEstimate sourceStats = lookup.getStats(semiJoinNode.getSource(), session, types);
+        PlanNodeStatsEstimate filteringSourceStats = lookup.getStats(semiJoinNode.getFilteringSource(), session, types);
+        Symbol filteringSourceJoinSymbol = semiJoinNode.getFilteringSourceJoinSymbol();
+        Symbol sourceJoinSymbol = semiJoinNode.getSourceJoinSymbol();
+
+        Optional<SemiJoinOutputFilter> semiJoinOutputFilter = extractSemiJoinOutputFilter(filterNode.getPredicate(), semiJoinNode.getSemiJoinOutput());
+
+        if (!semiJoinOutputFilter.isPresent()) {
+            return Optional.empty();
+        }
+
+        PlanNodeStatsEstimate semiJoinStats;
+        if (semiJoinOutputFilter.get().isNegated()) {
+            semiJoinStats = semiJoinStatsCalculator.computeAntiJoin(sourceStats, filteringSourceStats, sourceJoinSymbol, filteringSourceJoinSymbol);
+        }
+        else {
+            semiJoinStats = semiJoinStatsCalculator.computeSemiJoin(sourceStats, filteringSourceStats, sourceJoinSymbol, filteringSourceJoinSymbol);
+        }
+
+        // apply remaining prediate
+        PlanNodeStatsEstimate intermediateFilterStats = filterStatsCalculator.filterStats(semiJoinStats, semiJoinOutputFilter.get().getRemainingPredicate(), session, types);
+
+        // filter out unneeded symbols
+        PlanNodeStatsEstimate.Builder resultStats = PlanNodeStatsEstimate.buildFrom(intermediateFilterStats);
+        intermediateFilterStats.getSymbolsWithKnownStatistics().stream()
+                .filter(s -> !filterNode.getOutputSymbols().contains(s))
+                .forEach(s -> resultStats.removeSymbolStatistics(s));
+        return Optional.of(resultStats.build());
+    }
+
+    private Optional<SemiJoinOutputFilter> extractSemiJoinOutputFilter(Expression predicate, Symbol semiJoinOutput)
+    {
+        List<Expression> conjuncts = extractConjuncts(predicate);
+        List<Expression> semiJoinOutputReferences = conjuncts.stream()
+                .filter(conjunct -> isSemiJoinOutputReference(conjunct, semiJoinOutput))
+                .collect(toImmutableList());
+
+        if (semiJoinOutputReferences.size() != 1) {
+            return Optional.empty();
+        }
+
+        Expression semiJoinOutputReference = Iterables.getOnlyElement(semiJoinOutputReferences);
+        Expression remainingPredicate = combineConjuncts(conjuncts.stream()
+                .filter(conjunct -> conjunct != semiJoinOutputReference)
+                .collect(toImmutableList()));
+        return Optional.of(new SemiJoinOutputFilter(semiJoinOutputReference instanceof NotExpression, remainingPredicate));
```

Comment:
consider assign `semiJoinOutputReference instanceof NotExpression` to a variable